### DISCUSSION TITLE
How can I measure how long each preprocessing step takes?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [dita-ot/dita-ot#3995] [Q&A] [answered]
DITA-OT preprocessing consists of [various stages](https://www.dita-ot.org/dev/reference/preprocessing.html) (`gen-list`, `mapref`, `mappull`, `topicpull`, etc.). How do I measure the runtime used by each of these stages during a transformation?

**Comment by chrispy-snps** [comment, marked as answer]
To measure the runtime of each processing step of a DITA-OT transformation, you can do the following:

1. Enable Ant profiling by setting the following variable in your environment:

   ```
   # bash
   export ANT_ARGS="-logger org.apache.tools.ant.listener.ProfileLogger"

   # csh/tcsh
   setenv ANT_ARGS "-logger org.apache.tools.ant.listener.ProfileLogger"
   ```

2. Run your transformation and redirect the output to a log file:

   ```
   dita ... > out.log
   ```

3. Use the attached script to extract the per-target runtimes:

   ```
   extract_target_runtimes.pl out.log
   ```

   Script attachment: [extract_target_runtimes.zip](https://github.com/dita-ot/dita-ot/files/9477297/extract_target_runtimes.zip)

The script extracts the information added by the Ant profile logger, filters any targets that complete in less than one second, then prints the results as an indented list:

```
init started
 gen-list started
 gen-list finished - 33.669
 debug-filter started
 debug-filter finished - 78.929
 mapref started
 mapref finished - 5.854
 branch-filter started
 branch-filter finished - 3.417
 keyref started
 keyref finished - 80.417
 topic-fragment started
 topic-fragment finished - 24.909
 chunk started
 chunk finished - 1.422
 move-meta-entries started
 move-meta-entries finished - 60.194
 maplink started
 maplink finished - 40.255
 ...omitted...
 clean-temp started
 clean-temp finished - 20.283
init finished - 442.922
```

**Comment by markgif** [comment]
nice

**Comment by raducoravu** [comment]
Thanks for adding this discussion @chrispy-snps I just gave someone a link to this.
